### PR TITLE
Feat/optuna hpo 340 v3

### LINKED ISSUE #340: [Feature]: Optuna hyperparameter optimization integration
## Problem

Hyperparameter tuning (distortion strength schedule, learning rate, cycle count, convergence threshold) is currently manual. Users must guess values or run grid searches by hand. Best-in-class ML frameworks (Lightning, Ray Tune) integrate with Optuna for automated HPO.

## Proposed Solution

1. Add Optuna as an optional dependency (`nightmarenet[hpo]`)
2. Create a `HyperparameterOptimizer` class that wraps the pipeline runner
3. Define a search space from config (which params to tune, ranges)
4. Objective function: robustness score from evaluation phase
5. Support pruning (MedianPruner) to stop bad trials early

## Acceptance Criteria

- [ ] Optuna integration via `nightmarenet[hpo]` optional extra
- [ ] Search space configurable via YAML (e.g., `hpo.search_space` in config)
- [ ] Supports at least: learning rate, distortion strength, cycle count, batch size
- [ ] Pruning enabled (MedianPruner) to stop unpromising trials
- [ ] Results logged in Optuna study (SQLite or in-memory)
- [ ] CLI command: `nightmarenet optimize --config config.yaml --n-trials 50`
- [ ] Test verifies optimization loop runs with mock objective

## Scope

- `nightmarenet/optimization/hpo.py` - new module
- `nightmarenet/cli.py` - add `optimize` command
- `pyproject.toml` - add `[hpo]` optional extra
- `configs/examples/hpo-config.yaml` - example HPO config
- `tests/test_hpo.py` - new test file

> **Important:** Please read CONTRIBUTING.md in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

### PR DESCRIPTION
**Pull request by Ayush-0918**
## Summary

Add Optuna-based hyperparameter optimization support with a dedicated CLI command for automated hyperparameter tuning.

## Motivation

NightmareNet currently requires manual hyperparameter tuning, which is time-consuming and difficult to reproduce. This change introduces Optuna-based optimization to automate the search process and improve the training workflow.

Closes #340

## Changes

- Added Optuna-based hyperparameter optimization module.
- Added `optimize` CLI command for running HPO.
- Added configurable `--config` and `--n-trials` CLI options.
- Added graceful handling when Optuna is not installed.
- Added tests covering the new optimization functionality.
- Addressed previous review feedback and ensured compatibility with the current codebase.

## Acceptance Criteria

- [x] Add Optuna-based hyperparameter optimization support.
- [x] Provide a CLI command to execute optimization.
- [x] Allow configurable optimization trials.
- [x] Handle missing Optuna dependency gracefully.
- [x] Include tests for the new functionality.

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [ ] Refactor (no functional change)
- [ ] Documentation
- [x] Tests
- [ ] Infrastructure (CI/CD, Docker, tooling)

---

## Pre-submission Requirements

- [x] I have **starred** this repository ⭐
- [x] I have **followed** [@Adit-Jain-srm](https://github.com/Adit-Jain-srm)
- [x] I have read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full

## Quality Checklist

- [x] `ruff check nightmarenet/ scripts/ tests/` — 0 errors
- [ ] `mypy nightmarenet/ --ignore-missing-imports` — no new errors (run on Python 3.12)
- [x] `pytest tests/` — all tests pass
- [x] New functionality has corresponding tests
- [x] Commit messages follow Conventional Commits
- [x] No `from __future__ import annotations` added under `nightmarenet/api/`
- [x] No new imports of hosted-only libraries (`sqlalchemy`, `redis`, `celery`) in `nightmarenet/`

## Documentation

- [ ] No documentation needed (test-only or internal refactor)
- [ ] README.md updated
- [x] Docstrings added/updated (Google style, public APIs only)
- [ ] `docs/` updated (architecture, research, or API docs)
- [ ] Config comments updated (`configs/default.yaml`)
- [ ] CHANGELOG entry added (if user-facing change)

## AI Disclosure

- [x] This PR includes AI-assisted code (Copilot, ChatGPT, Claude, etc.) — I have reviewed every line and can explain it

## Security Considerations

- [x] Not applicable (no security-sensitive changes)
- [x] User input is validated before use
- [x] No secrets, keys, or credentials in code or comments
- [x] CORS / auth / rate limiting behavior unchanged (or explicitly documented)
- [ ] If this is a security fix, see SECURITY.md

## Screenshots (if UI/UX change)

N/A

## Breaking Changes

N/A

---

<details>
<summary>Reviewer notes (optional)</summary>

This is a fresh submission of the Optuna HPO feature after the previous PR was reverted due to an ECSOC spam flag. The implementation has been restored, conflicts resolved against the latest `main`, and all previous review feedback has been incorporated. Local validation (`ruff` and `pytest`) passes successfully.
</details>

## Summary by CodeRabbit

* **New Features**
  * Added hyperparameter optimization via an Optuna-powered workflow configured through a new `hpo` section in YAML.
  * Introduced an `optimize` CLI command (requires `--config`, supports optional `--n-trials`) with study persistence and pruning behavior.
  * Added a sample HPO configuration example.
* **Bug Fixes**
  * Improved failure handling with clearer messaging when optional optimization tooling is unavailable or when optimization can’t run.
* **Tests**
  * Added coverage for Optuna availability checks, config loading, and trial execution behavior.

### CODE REVIEW

File: nightmarenet/cli.py [hunk 1]
```diff
@@ -822,6 +853,17 @@ def build_parser() -> argparse.ArgumentParser:
...
+    optimize_parser = subparsers.add_parser(
+        "optimize", help="Run hyperparameter optimization via Optuna"
+    )
+    optimize_parser.add_argument("--config", required=True, help="YAML config path")
+    optimize_parser.add_argument(
+        "--n-trials",
+        type=int,
+        help="Override the number of optimization trials.",
+    )
+
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply nightmarenet/cli.py:866]: Thanks for the suggestion! I’ve initialized pipeline to None before defining the callback so it can be safely referenced even if events are emitted during initialization.

File: nightmarenet/optimization/hpo.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                current_cycle = metrics.get("current_cycle", 0)
+
+                if current_cycle > 0:
+                    trial.report(phase_loss, step=current_cycle)
+                    if trial.should_prune():
+                        pruned_flag[0] = True
+                        if pipeline is not None:
+                            pipeline.cancel()
+
+        pipeline = Pipeline(config=trial_config, on_event=on_event)
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply nightmarenet/optimization/hpo.py:124]: Good catch! I’ve added mean_robustness to the fallback metric list to keep the behavior consistent with Pipeline.evaluate() and avoid falling back to 0.0 when that metric is returned.

File: nightmarenet/optimization/hpo.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        if pruned_flag[0] or pipeline._context.cancelled:
+            raise optuna.exceptions.TrialPruned()
+
+        metric_val = comparison.get("robustness_delta")
+        if metric_val is None:
+            for fallback in ("robustness", "avg_robustness"):
+                if fallback in comparison:
+                    metric_val = comparison[fallback]
+                    break
```
[the remark this answers is not in the document]
Comment (Ayush-0918) [reply nightmarenet/optimization/hpo.py]: Thanks for the review! I’ve addressed the suggested improvements, verified the changes locally, and pushed the updated implementation.

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
